src/service/resources-impl.js

```diff
@@ -190,15 +193,51 @@ export class ResourcesImpl {
       this.ampdoc.getVisibilityState()
     );
 
-    // When viewport is resized, we have to re-measure all elements.
+    if (isExperimentOn(this.win, 'intersect-resources')) {
+      const platform = Services.platformFor(this.win);
+      // As of 1/2020, only Chrome 81+ has the required behavior for using
+      // rootMargin with viewport tracking in a cross-origin iframe (#25428).
+      const supportsIframes =
+        platform.isChrome() && platform.getMajorVersion() >= 81;
+      const iframed = isIframed(this.win);
+      devAssert(supportsIframes || !iframed);
+
+      // Need to use scrollingElement as root for viewport tracking in iframed pages.
+      const root =
+        this.ampdoc.isSingleDoc() && iframed
+          ? this.win.document./*OK*/ scrollingElement
+          : null;
+      // TODO(willchou): Is 3x viewport loading rectangle too large given that
+      // IntersectionObserver is more responsive than scroll-bound measure?
+      // TODO(willchou): Support prerenderSize_ loading rectangle.
+      const rootMargin = '200% 25%';
+      /** @private @const {?IntersectionObserver} */
+      this.intersectionObserver_ = new IntersectionObserver(
+        this.intersects_.bind(this),
+        {
+          root,
+          rootMargin,
+        }
+      );
+
+      // Wait for intersection callback instead of measuring all elements
+      // during the first pass.
+      this.relayoutAll_ = false;
+    }
+
+    // When user scrolling stops, run pass to check newly in-viewport elements.
+    // When viewport is resized, we have to re-measure everything.
     this.viewport_.onChanged(event => {
       this.lastScrollTime_ = Date.now();
       this.lastVelocity_ = event.velocity;
       if (event.relayoutAll) {
         this.relayoutAll_ = true;
         this.maybeChangeHeight_ = true;
       }
-      this.schedulePass();
+      // With IntersectionObserver, we only need to handle viewport resize.
+      if (this.relayoutAll_ || !this.intersectionObserver_) {
+        this.schedulePass();
```

Comment:
We do use `schedulePass` for build at this time. Would be nice to get rid of that, but it'd be a pretty big undertaking.